In [1]:
import pandas as pd
import numpy as np

In [2]:
data =  pd.read_csv('Dataset_superstore.csv', parse_dates=['order_date']) 
# parse date membuat kolom order_date menjadi tipe date
data.head()
#len(data.order_id.unique())
#len(data.customer_id.unique())

order_id order_date customer_id    segment         category  \
0  CA-2016-152156 2016-11-08    CG-12520   Consumer        Furniture   
1  CA-2016-152156 2016-11-08    CG-12520   Consumer        Furniture   
2  CA-2016-138688 2016-06-12    DV-13045  Corporate  Office Supplies   
3  US-2015-108966 2015-10-11    SO-20335   Consumer        Furniture   
4  US-2015-108966 2015-10-11    SO-20335   Consumer  Office Supplies   

  sub_category     sales  quantity    profit  
0    Bookcases  261.9600         2   41.9136  
1       Chairs  731.9400         3  219.5820  
2       Labels   14.6200         2    6.8714  
3       Tables  957.5775         5 -383.0310  
4      Storage   22.3680         2    2.5164

In [3]:
data.dtypes

order_id                object
order_date      datetime64[ns]
customer_id             object
segment                 object
category                object
sub_category            object
sales                  float64
quantity                 int64
profit                 float64
dtype: object

#### <p style='color:red'>Akan dicari customer_id yang memiliki sales paling besar</p>

In [4]:
# dicari dengan men-sort kolom sales
data.sort_values('sales', ascending=False)

order_id order_date customer_id    segment         category  \
9033  CA-2016-117121 2016-12-17    AB-10105   Consumer  Office Supplies   
4096  CA-2014-116904 2014-09-23    SC-20095   Consumer  Office Supplies   
4274  US-2016-107440 2016-04-16    BS-11365  Corporate       Technology   
8482  CA-2016-158841 2016-02-02    SE-20110   Consumer       Technology   
6422  CA-2016-143714 2016-05-23    CC-12370   Consumer       Technology   
...              ...        ...         ...        ...              ...   
2106  US-2014-152723 2014-09-26    HG-14965  Corporate  Office Supplies   
4708  CA-2014-112403 2014-03-31    JO-15280   Consumer  Office Supplies   
8652  CA-2016-168361 2016-06-21    KB-16600  Corporate  Office Supplies   
9286  CA-2017-124114 2017-03-02    RS-19765  Corporate  Office Supplies   
4099  US-2017-102288 2017-06-19    ZC-21910   Consumer  Office Supplies   

     sub_category     sales  quantity     profit  
9033      Binders  9892.740        13  4946.3700  
4096      Binders  9449.950         5  4630.4755  
4274     Machines  9099.930         7  2365.9818  
8482     Machines  8749.950         5  2799.9840  
6422      Copiers  8399.976         4  1119.9968  
...           ...       ...       ...        ...  
2106      Binders     0.876         1    -1.4016  
4708      Binders     0.852         1    -0.5964  
8652      Binders     0.836         1    -1.3376  
9286      Binders     0.556         1    -0.9452  
4099   Appliances     0.444         1    -1.1100  

[9988 rows x 9 columns]

<pre>Didapat costumer_id yg memiliki 'sales' paling besar adalah <p style='color:red'>AB-10105</p>dan berikut informasi lainnya dapat dilihat pada DataFrame diatas</pre>

#### <p style='color:red'>Banyak order yang menghasilkan profit negatif (rugi)</p>

In [5]:
data2 = data.sort_values('profit', ascending=True)
data2=data2.reset_index()
data2

index        order_id order_date customer_id      segment  \
0       683  US-2017-168116 2017-11-04    GT-14635    Corporate   
1      9768  CA-2014-169019 2014-07-26    LF-17185     Consumer   
2      3009  CA-2017-134845 2017-04-17    SR-20425  Home Office   
3      4988  US-2017-122714 2017-12-07    HG-14965    Corporate   
4      3149  CA-2015-147830 2015-12-15    NF-18385     Consumer   
...     ...             ...        ...         ...          ...   
9983   7662  US-2016-140158 2016-10-04    DR-12940  Home Office   
9984   8482  CA-2016-158841 2016-02-02    SE-20110     Consumer   
9985    509  CA-2015-145352 2015-03-16    CM-12385     Consumer   
9986   4096  CA-2014-116904 2014-09-23    SC-20095     Consumer   
9987   9033  CA-2016-117121 2016-12-17    AB-10105     Consumer   

             category sub_category     sales  quantity     profit  
0          Technology     Machines  7999.980         4 -3839.9904  
1     Office Supplies      Binders  2177.584         8 -3701.8928  
2          Technology     Machines  2549.985         5 -3399.9800  
3     Office Supplies      Binders  1889.990         5 -2929.4845  
4          Technology     Machines  1799.994         2 -2639.9912  
...               ...          ...       ...       ...        ...  
9983       Technology      Copiers  5399.910         9  2591.9568  
9984       Technology     Machines  8749.950         5  2799.9840  
9985  Office Supplies      Binders  6354.950         5  3177.4750  
9986  Office Supplies      Binders  9449.950         5  4630.4755  
9987  Office Supplies      Binders  9892.740        13  4946.3700  

[9988 rows x 10 columns]

In [6]:
profit=np.array(data2['profit'])
a=[]
def rugi(x):
    for i in profit:
        if i<0:
            a.append(i)

rugi(profit)
print(len(a))
print(sum(a))
#print(a)

1869
-147720.2292999998


banyak order yang menghasilkan profit negatif (rugi) per sub_category sebanyak : <b>1869</b><br>
dengan total kerugian : <b>-147720.2292999998</b>


#### <p style="color:red">Macam sub_category dalam category 'office_supplies' dan total profit masing-masing sub_category nya</p>

In [7]:
data3 = data.groupby(['category','sub_category']).agg({
    'profit':'sum'
})
data3

profit
category        sub_category            
Furniture       Bookcases     -3472.5560
                Chairs        26590.1663
                Furnishings   13059.1436
                Tables       -17725.4811
Office Supplies Appliances    18138.0054
                Art            6527.7870
                Binders       30221.7633
                Envelopes      6964.1767
                Fasteners       949.5182
                Labels         5546.2540
                Paper         34053.5693
                Storage       21278.8264
                Supplies      -1189.0995
Technology      Accessories   41936.6357
                Copiers       31537.8937
                Machines      11795.8133
                Phones        44515.7306

#### Cek manakah diantara 3 customer KH-16510, AD-10180,  JE-16165 yang total salesnya paling banyak
sebelum itu, kita dapat lihat data frame diatas per customer dapat menjadi beberapa baris <br>
sehingga saya akan mensort per-customer 1 baris saja

In [8]:
data4 = data.groupby(['customer_id']).agg({
    'sales':'sum',
    'profit':'sum'
})
data4 = data4.reset_index()
data4 = data4.sort_values('customer_id')
data4

customer_id     sales     profit
0      AA-10315  5563.560  -362.8825
1      AA-10375  1056.390   277.3824
2      AA-10480  1790.512   435.8274
3      AA-10645  5086.935   857.8033
4      AB-10015   886.156   129.3465
..          ...       ...        ...
788    XP-21865  2374.658   621.2300
789    YC-21895  5454.350  1305.6290
790    YS-21880  6720.444  1778.2923
791    ZC-21910  8025.707 -1032.1490
792    ZD-21925  1493.944   249.1307

[793 rows x 3 columns]

In [9]:
cust = np.array(data4['customer_id'])
b=[]
for i in range (0,len(cust)):
    if cust[i]=='KH-16510':
        print('index KH-16510 :',i)
        b.append(i)
    elif cust[i]=='AD-10180':
        print('index AD-10180 :',i)
        b.append(i)
    elif cust[i]=='JE-16165':
        print('index JE-16165 :',i)
        b.append(i)

print('\n',b)

index AD-10180 : 15
index JE-16165 : 357
index KH-16510 : 430

 [15, 357, 430]


In [10]:
sl=np.array(data4['sales'])
cc=np.array(data4['customer_id'])

customer=[cc[15], cc[357], cc[430]]
sales=[sl[15], sl[357], sl[430]]

a1=pd.DataFrame({
    'customer':customer,
    'sales':sales
})
a1
# jadi, nilai total sales terbesar yaitu customer AD-10180

customer     sales
0  AD-10180  6106.880
1  JE-16165  2697.248
2  KH-16510  5952.859

### Analisis data per tahun
- total sales
- jumlah customer
- total profit<br>

tiap tahun semua

In [11]:
data['order_year']=data.order_date.dt.strftime('%Y')
data[:]

order_id order_date customer_id    segment         category  \
0     CA-2016-152156 2016-11-08    CG-12520   Consumer        Furniture   
1     CA-2016-152156 2016-11-08    CG-12520   Consumer        Furniture   
2     CA-2016-138688 2016-06-12    DV-13045  Corporate  Office Supplies   
3     US-2015-108966 2015-10-11    SO-20335   Consumer        Furniture   
4     US-2015-108966 2015-10-11    SO-20335   Consumer  Office Supplies   
...              ...        ...         ...        ...              ...   
9983  CA-2014-110422 2014-01-21    TB-21400   Consumer        Furniture   
9984  CA-2017-121258 2017-02-26    DB-13060   Consumer        Furniture   
9985  CA-2017-121258 2017-02-26    DB-13060   Consumer       Technology   
9986  CA-2017-121258 2017-02-26    DB-13060   Consumer  Office Supplies   
9987  CA-2017-119914 2017-05-04    CC-12220   Consumer  Office Supplies   

     sub_category     sales  quantity    profit order_year  
0       Bookcases  261.9600         2   41.9136       2016  
1          Chairs  731.9400         3  219.5820       2016  
2          Labels   14.6200         2    6.8714       2016  
3          Tables  957.5775         5 -383.0310       2015  
4         Storage   22.3680         2    2.5164       2015  
...           ...       ...       ...       ...        ...  
9983  Furnishings   25.2480         3    4.1028       2014  
9984  Furnishings   91.9600         2   15.6332       2017  
9985       Phones  258.5760         2   19.3932       2017  
9986        Paper   29.6000         4   13.3200       2017  
9987   Appliances  243.1600         2   72.9480       2017  

[9988 rows x 10 columns]

<p style='color:blue'>Dapat dilihat bahwa dalam data tsb customer_id masih banyak yang dobel, jadi perlu untuk dijadikan 1 baris hanya 1 customer_id yang berbeda (tiap tahunnya)

In [12]:
data5= data.groupby(['order_year','customer_id']).agg({
    'sales':'sum',
    'profit':'sum'
})

data5 = data5.sort_values('order_year')
data5 =data5.reset_index()
data5
#kalau len(data5.customer_id.unique()) saja ada = 793 
#tapi hasil groupby ada 2499 baris soalnya tiap tahun juga keitung, jadi customer_id bisa dobel

order_year customer_id     sales    profit
0          2014    AA-10315   756.048  280.7050
1          2014    MP-17470  1211.090  279.8571
2          2014    MP-18175  1724.460 -125.3022
3          2014    MR-17545   137.600    9.5920
4          2014    MS-17365   425.688  143.9010
...         ...         ...       ...       ...
2494       2017    EJ-13720  1219.792  423.7296
2495       2017    EJ-14155   513.290  140.0081
2496       2017    EK-13795    23.360    6.0736
2497       2017    EG-13900   468.972    2.5856
2498       2017    ZD-21925    61.440   16.5888

[2499 rows x 4 columns]

In [13]:
data5.order_year.unique()

array(['2014', '2015', '2016', '2017'], dtype=object)

In [14]:
tahun=np.array(data5['order_year'])
a,b,c,d=0,0,0,0
for i in range (0,len(tahun)):
    if tahun[i]=='2014':
        a=a+1
    elif tahun[i]=='2015':
        b=b+1
    elif tahun[i]=='2016':
        c=c+1
    elif tahun[i]=='2017':
        d=d+1
print(a,b,c,d)
print(f'2014, index 0 sampai ke {a-1}')  
print(f'2015, index {a} sampai ke {b+a-1}')
print(f'2016, index {b+a} sampai ke {c+b+a-1}')
print(f'2017, index {c+b+a} sampai ke {d+c+b+a-1}')

595 573 638 693
2014, index 0 sampai ke 594
2015, index 595 sampai ke 1167
2016, index 1168 sampai ke 1805
2017, index 1806 sampai ke 2498


<p style="color:red">Jumlah customer, sales, dan profit per tahun

In [15]:
print('jumlah costomer tahun 2014 =',a)
print('jumlah costomer tahun 2015 =',b)
print('jumlah costomer tahun 2016 =',c)
print('jumlah costomer tahun 2017 =',d)
cst=[a,b,c,d]

jumlah costomer tahun 2014 = 595
jumlah costomer tahun 2015 = 573
jumlah costomer tahun 2016 = 638
jumlah costomer tahun 2017 = 693


In [16]:
data6 = data.groupby(['order_year']).agg({
    'sales':'sum',
    'profit':'sum'
}) 
data6 = data6.reset_index()
data6['customer']=cst
data6

order_year        sales      profit  customer
0       2014  461609.0181  51355.0525       595
1       2015  470532.5090  61618.6037       573
2       2016  587205.6630  79995.1763       638
3       2017  697515.3572  77759.3144       693

jadi profit tertinggi diperoleh pada tahun 2016

In [17]:
#cek total cust disamakan DataFrame data5
tes=np.array(data6['customer'])
tes.sum()

2499

# Membuat visualisasi

In [18]:
import plotly
import plotly.graph_objs as go
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

init_notebook_mode(connected=True)

In [22]:
data = data.sort_values('order_year')
data = data.reset_index()
data

index        order_id order_date customer_id      segment  \
0      9584  CA-2014-119172 2014-05-11    HD-14785  Home Office   
1      7428  CA-2014-161249 2014-08-09    RD-19720     Consumer   
2      4901  US-2014-161613 2014-12-01    MC-17605    Corporate   
3      8045  CA-2014-144974 2014-09-26    CM-12715    Corporate   
4      4992  CA-2014-132227 2014-11-04    SZ-20035  Home Office   
...     ...             ...        ...         ...          ...   
9983   2328  CA-2017-138422 2017-09-23    KN-16705  Home Office   
9984   2320  CA-2017-165491 2017-03-20    HW-14935    Corporate   
9985   2318  CA-2017-122035 2017-07-20    EM-13825    Corporate   
9986   2351  CA-2017-109701 2017-12-02    AM-10360    Corporate   
9987   9987  CA-2017-119914 2017-05-04    CC-12220     Consumer   

             category sub_category    sales  quantity    profit order_year  
0     Office Supplies      Binders  104.580         9 -172.5570       2014  
1     Office Supplies    Fasteners    4.464         3   -0.9486       2014  
2           Furniture       Chairs  674.058         3  -19.2588       2014  
3           Furniture  Furnishings   21.184         1    4.7664       2014  
4     Office Supplies      Binders   52.064         4   18.8732       2014  
...               ...          ...      ...       ...       ...        ...  
9983  Office Supplies    Envelopes   14.352         3    5.2026       2017  
9984       Technology  Accessories  265.930         7   63.8232       2017  
9985       Technology  Accessories  389.970         3  132.5898       2017  
9986  Office Supplies   Appliances   79.470         3   22.2516       2017  
9987  Office Supplies   Appliances  243.160         2   72.9480       2017  

[9988 rows x 11 columns]

<p style='color:blue'>Misal saya akan membuat 'Sales vs Profit' hanya untuk tahun 2014 dan 2015 saja

In [40]:
data[4092:4096]
#data tahun 2014-2015 index [0:4094]  

index        order_id order_date customer_id    segment  \
4092   3487  CA-2015-157322 2015-07-02    RH-19600   Consumer   
4093   3488  CA-2015-157322 2015-07-02    RH-19600   Consumer   
4094   1289  CA-2016-154711 2016-11-22    TB-21355  Corporate   
4095   9625  CA-2016-168844 2016-05-31    NM-18520   Consumer   

             category sub_category    sales  quantity   profit order_year  
4092        Furniture       Chairs  382.116         6 -92.7996       2015  
4093  Office Supplies      Storage   68.600         5   6.0025       2015  
4094  Office Supplies          Art   20.820         3   7.4952       2016  
4095       Technology       Phones  179.700         6  88.0530       2016

In [43]:
data7 = data[:4094]
data7

index        order_id order_date customer_id      segment  \
0      9584  CA-2014-119172 2014-05-11    HD-14785  Home Office   
1      7428  CA-2014-161249 2014-08-09    RD-19720     Consumer   
2      4901  US-2014-161613 2014-12-01    MC-17605    Corporate   
3      8045  CA-2014-144974 2014-09-26    CM-12715    Corporate   
4      4992  CA-2014-132227 2014-11-04    SZ-20035  Home Office   
...     ...             ...        ...         ...          ...   
4089   5712  CA-2015-124107 2015-10-09    BM-11650    Corporate   
4090   5708  US-2015-160857 2015-05-08    NW-18400     Consumer   
4091   5707  US-2015-160857 2015-05-08    NW-18400     Consumer   
4092   3487  CA-2015-157322 2015-07-02    RH-19600     Consumer   
4093   3488  CA-2015-157322 2015-07-02    RH-19600     Consumer   

             category sub_category     sales  quantity     profit order_year  
0     Office Supplies      Binders   104.580         9  -172.5570       2014  
1     Office Supplies    Fasteners     4.464         3    -0.9486       2014  
2           Furniture       Chairs   674.058         3   -19.2588       2014  
3           Furniture  Furnishings    21.184         1     4.7664       2014  
4     Office Supplies      Binders    52.064         4    18.8732       2014  
...               ...          ...       ...       ...        ...        ...  
4089       Technology  Accessories   619.950         5   111.5910       2015  
4090  Office Supplies        Paper    19.440         3     9.3312       2015  
4091       Technology      Copiers  2799.944         7  1014.9797       2015  
4092        Furniture       Chairs   382.116         6   -92.7996       2015  
4093  Office Supplies      Storage    68.600         5     6.0025       2015  

[4094 rows x 11 columns]

In [46]:
from scipy import stats
slope, intercept, r_value, p_value, std_err = stats.linregress(x=data7.sales, y=data7.profit)
lines = slope * data7.sales.values + intercept
lines

array([  7.12560236,  -9.52125216, 101.81593581, ..., 455.29904527,
        53.27308547,   1.14300392])

In [48]:
dot1 = go.Scatter(
    x=data7[data7.category=='Furniture'].sales,
    y=data7[data7.category=='Furniture'].profit,
    mode='markers',
    marker={
        'size':10
    },
    name='Furniture'
)

dot2 = go.Scatter(
    x=data7[data7.category=='Office Supplies'].sales,
    y=data7[data7.category=='Office Supplies'].profit,
    mode='markers',
    marker={
        'size':5
    },
    name='Office Supplies'
)

dot3 = go.Scatter(
    x=data7[data7.category=='Technology'].sales,
    y=data7[data7.category=='Technology'].profit,
    mode='markers',
    name='Technology'
)

line= go.Scatter(
    x=data7.sales,
    y=lines,
    line={
        'dash':'dash',
        'width': 5
    },
    name='Linear Trend Line'
)


data=[dot1, dot2, dot3, line]

layout = {
    'title': {
        'text' : 'Sales VS Profit 2014-2015',
        'x' : 0.5 # nilai x antara 0 - 1, jadi 0.5 agar title ditengah
    },
    
    'xaxis': { # memberi label
        'title':'Sales',
        'range' : [0,7500] # membatasi nilai 
    },
    'yaxis': {
        'title':'Profit',
        'range' : [-2500,2500]
        
    },
    
    'legend': {  # mengubah posisi legend 
        'orientation' : 'h', #jadi horizontal dibawah
        'x':0.5,
        # 0,5 kiri ;  1 kanan  (atas)
        'y':1,#jadi horizontal diatas
        'xanchor':'center',#horizontal diatas tengah
    }
    
}

fig=go.Figure(data=data,layout=layout)
iplot(fig)

<p style="color:blue"> Bar Chart total profit dari 10 customer dengan total sales tertinggi di tahun 2105

In [59]:
#dari data5 tahun 2015 mulai index [595:1168]
data8 = data5[595:1168]
data8 = data8.sort_values('sales', ascending=False).reset_index()
data8
#diambil 10 teratas
data8 =  data8[:10]
data8

index order_year customer_id     sales     profit
0    642       2015    PF-19120  9022.324  -631.9789
1    670       2015    KD-16495  7039.044  2810.0215
2    941       2015    CM-12385  6682.260  3197.4580
3   1141       2015    FH-14365  6056.090  2139.7363
4    633       2015    NW-18400  5511.316  1654.2153
5    877       2015    SB-20185  5414.246   939.1590
6   1094       2015    JH-15985  5352.786 -1429.3776
7   1027       2015    HM-14860  5327.802  2450.9987
8    883       2015    CB-12025  4662.553   223.4447
9    842       2015    QJ-19255  4409.850  1014.4635

In [62]:
data = go.Bar(
    x=data8.customer_id, 
    y=data8.sales
)

layout = {
    'title': {
        'text' : 'Customer id vs Sales',
        'x' : 0.5 # nilai x antara 0 - 1, jadi 0.5 agar title ditengah
    },
    
    'xaxis': { # memberi label
        'title':'Customer id'
    },
    'yaxis': {
        'title':'Sales'
        
    }
}
    
fig=go.Figure(data=data, layout=layout)
iplot(fig)

In [63]:
data = go.Bar(
    x=data8.customer_id, 
    y=data8.profit
)

layout = {
    'title': {
        'text' : 'Customer id vs profit',
        'x' : 0.5 # nilai x antara 0 - 1, jadi 0.5 agar title ditengah
    },
    
    'xaxis': { # memberi label
        'title':'Customer id'
    },
    'yaxis': {
        'title':'Profit'
        
    }
}
    
fig=go.Figure(data=data, layout=layout)
iplot(fig)

Dapat dilihat bahwa pada 10 customer dengan sales tertinggi ada beberapa customer yang justru menghasilkan profit minus pada perusahaan<br>